### PR TITLE
`MusicUtils` to MediaManager

### PR DESCRIPTION
**Pull request by fxsalazar**
I just did the abstraction to the interface `MediaManager`; 
Small changes on the DI classes to make it all fit.

### CODE REVIEW

**Comment by timusus** [comment]
Thanks, I was going to ask you to change the `musicUtils` vars to `mediaManager`.

I'm not sure what's going on with the imports. I think I have some weird settings in my local environment which have messed up imports project-wide, so I'll see if I can address that, which should clean up future PRs.

File: app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt [hunk 1]
```diff
@@ -44,27 +44,21 @@ class AlbumDetailPresenter constructor(private val album: Album) : Presenter<Alb
     }
 
     fun fabClicked() {
-        MusicUtils.shuffleAll(songs) { message ->
-            view?.showToast(message)
-        }
+        musicUtils.shuffleAll(songs, { view?.showToast(it) })
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt]: Although this is more concise, the convention in Shuttle is to prefer not to use implicit naming of lambda parameters. (Prefer `message` to `it`).

File: app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt [hunk 2]
```diff
@@ -44,27 +44,21 @@ class AlbumDetailPresenter constructor(private val album: Album) : Presenter<Alb
...
     }
 
     fun playAll() {
-        MusicUtils.playAll(songs, 0, true) { message ->
+        musicUtils.playAll(songs, 0, true) { message ->
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt]: Prefer explicit parameter naming in lambdas. (Prefer not to use `it`).

File: app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt [hunk 3]
```diff
@@ -44,27 +44,21 @@ class AlbumDetailPresenter constructor(private val album: Album) : Presenter<Alb
...
             view?.showToast(message)
         }
     }
 
     fun playNext() {
-        MusicUtils.playNext(songs) { message ->
-            view?.showToast(message)
-        }
+        musicUtils.playNext(songs) { view?.showToast(it) }
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt]: Prefer explicit parameter naming in lambdas. (Prefer not to use `it`).

File: app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt [hunk 4]
```diff
@@ -44,27 +44,21 @@ class AlbumDetailPresenter constructor(private val album: Album) : Presenter<Alb
...
     }
 
     fun addToQueue() {
-        MusicUtils.addToQueue(songs) { message ->
-            view?.showToast(message)
-        }
+        musicUtils.addToQueue(songs) { view?.showToast(it) }
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt]: Prefer explicit parameter naming in lambdas. (Prefer not to use `it`).

File: app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt [hunk 5]
```diff
@@ -93,8 +87,6 @@ class AlbumDetailPresenter constructor(private val album: Album) : Presenter<Alb
     }
 
     fun songClicked(song: Song) {
-        MusicUtils.playAll(songs, songs.indexOf(song), true) { message ->
-            view?.showToast(message)
-        }
+        musicUtils.playAll(songs, songs.indexOf(song), true) { view?.showToast(it) }
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/album/AlbumDetailPresenter.kt]: Prefer explicit parameter naming in lambdas. (Prefer not to use `it`).

File: app/src/main/java/com/simplecity/amp_library/ui/detail/genre/GenreDetailPresenter.kt [hunk 6]
```diff
@@ -84,7 +85,7 @@ class GenreDetailPresenter constructor(private val genre: Genre) : Presenter<Gen
             .delay(if (currentSlideShowAlbum == null) 0L else 8L, TimeUnit.SECONDS)
 
         addDisposable(Observable
-            .combineLatest(a, b, BiFunction { albums: List<Album>, aLong: Long -> albums })
+            .combineLatest(a, b, BiFunction { albums: List<Album>, _: Long -> albums })
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/genre/GenreDetailPresenter.kt]: There's nothing wrong with this change, but it probably should be part of a separate commit, since it has nothing to do with `MusicUtils`

File: app/src/main/java/com/simplecity/amp_library/ui/detail/playlist/PlaylistDetailPresenter.kt [hunk 7]
```diff
@@ -12,15 +13,14 @@ import io.reactivex.schedulers.Schedulers
 import java.util.*
 import java.util.concurrent.TimeUnit
 
-class PlaylistDetailPresenter constructor(private val playlist: Playlist) : Presenter<PlaylistDetailView>() {
+class PlaylistDetailPresenter @JvmOverloads constructor(private val playlist: Playlist, val musicUtils: MediaManager = MusicUtils()) : Presenter<PlaylistDetailView>() {
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/playlist/PlaylistDetailPresenter.kt]: Why are we using `JVMOverloads` here?

File: app/src/main/java/com/simplecity/amp_library/ui/detail/playlist/PlaylistDetailPresenter.kt [hunk 8]
```diff
@@ -97,25 +97,25 @@ class PlaylistDetailPresenter constructor(private val playlist: Playlist) : Pres
     }
 
     fun fabClicked() {
-        MusicUtils.shuffleAll(songs) { message ->
+        musicUtils?.shuffleAll(songs) { message ->
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/detail/playlist/PlaylistDetailPresenter.kt]: Why is this optional?

File: app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/CustomMultiSheetView.java [hunk 9]
```diff
@@ -22,8 +17,8 @@
 
     private static final String TAG = "CustomMultiSheetView";
 
-    @Inject MultiSheetEventRelay multiSheetEventRelay;
-    @Inject MultiSheetSlideEventRelay multiSheetSlideEventRelay;
+    MultiSheetEventRelay multiSheetEventRelay = new MultiSheetEventRelay();
+    MultiSheetSlideEventRelay multiSheetSlideEventRelay = new MultiSheetSlideEventRelay();
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/CustomMultiSheetView.java]: We should not be instantiating the `MultiSheetSlideEventRelay()` here. Since it's an event really, there should only be one instance across the app. This is instantiated in the `MultiSheetModule` in order to inject it elsewhere.

File: app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/CustomMultiSheetView.java [hunk 10]
```diff
@@ -22,8 +17,8 @@
 
     private static final String TAG = "CustomMultiSheetView";
 
-    @Inject MultiSheetEventRelay multiSheetEventRelay;
-    @Inject MultiSheetSlideEventRelay multiSheetSlideEventRelay;
+    MultiSheetEventRelay multiSheetEventRelay = new MultiSheetEventRelay();
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/CustomMultiSheetView.java]: If we're going to instantiate the `MultiSheetEventRelay` here, we need to make sure we're not instantiating a separate instance elsewhere (which we are doing in `MultiSheedModule`).

File: app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetSlideEventRelay.java [hunk 11]
```diff
@@ -1,19 +1,14 @@
...
-
-import javax.inject.Inject;
-
 import io.reactivex.Observable;
 
 public class MultiSheetSlideEventRelay {
 
     private final BehaviorRelay<SlideEvent> eventRelay = BehaviorRelay.create();
 
-    @Inject
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetSlideEventRelay.java]: What's the purpose of this change?

File: app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetEventRelay.java [hunk 12]
```diff
@@ -3,16 +3,12 @@
...
-
-import javax.inject.Inject;
-
 import io.reactivex.Observable;
 
 public class MultiSheetEventRelay {
 
     private PublishRelay<MultiSheetEvent> eventRelay = PublishRelay.create();
 
-    @Inject
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetEventRelay.java]: What's the purpose of this change?

File: app/src/main/java/com/simplecity/amp_library/ui/views/UpNextView.java [hunk 13]
```diff
@@ -90,15 +80,21 @@
 
     private CompositeDisposable disposables = new CompositeDisposable();
 
-    public UpNextView(Context context) {
+    public static UpNextView newInstance(Context context, PlayerPresenter playerPresenter){
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/views/UpNextView.java]: Why are we making this `View` a singleton? 

This leaks  a `Context`.

File: app/src/main/java/com/simplecity/amp_library/playback/MediaManager.kt [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+import com.simplecity.amp_library.model.Song
+import io.reactivex.Single
+
+/**
+ * Created by fxsalazar
+ * 01/05/2018.
+ */
+interface MediaManager {
+
+    interface Defs {
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/playback/MediaManager.kt]: This appears to be re-declared in the `MusicUtils` implementation of `MediaManager`

File: app/src/main/java/com/simplecity/amp_library/playback/MediaManager.kt [hunk 15]
```diff
@@ -0,0 +1,164 @@
+package com.simplecity.amp_library.playback
+
+import android.net.Uri
+import com.simplecity.amp_library.model.Album
+import com.simplecity.amp_library.model.AlbumArtist
+import com.simplecity.amp_library.model.Genre
+import com.simplecity.amp_library.model.Song
+import io.reactivex.Single
+
+/**
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/playback/MediaManager.kt]: Author/date headers are not used throughout Shuttle. Authorship is maintained through version control.

**Comment by fxsalazar** [comment]
I think I covered all of the requested changes.

Following your comments:
> I'm not entirely convinced that this makes MusicUtils less of a 'God Object' - it still has the same scope, but the functions are no longer static.
Now `MusicUtils` is a 'God Implementation' of `MediaManager` which eventually will allow to plug the other implementation of the player. As for the scope, I just wanted to do it little by little to not to have huge PRs.

> I've added some comments on a few changes which appear unrelated to MusicUtils refactoring. In future, it would be good to keep unrelated changes in their own commits / PR's - to make sure we don't introduce bugs that are hard to track down.
I totally agree with this; I just like to remove/clean/update _harmless_ warnings/format/readability when possible. ex. the `aLong`, `message` for `it` because I saw that you were using it in some cases.

> It looks like there are some differences between our import ordering strategies. I'll see if I can reset mine to default, and reformat the project, to make future merges easier.
Is because I'm using the **optimise imports on the fly**; it just helps considerably the importing process as well as reducing lines on the file.
screen shot 2018-05-05 at 10 59 48

File: app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetSlideEventRelay.java (hunk 11 above)
Comment (fxsalazar) [reply app/src/main/java/com/simplecity/amp_library/ui/views/multisheet/MultiSheetSlideEventRelay.java]: I wanted to start flattening the DI structured, but you are right, better to do it separately. I just put the Injection back but removed the unnecessary Module.

**Comment by fxsalazar** [comment]
> This shouldn't be a problem - I also use 'optimise imports on the fly'. The issue is that our imports are not optimised in the same fashion. This should be resolved now that there is a project level code style.

BTW, I think you forgot to commit `.idea/codeStyleSettings.xml` because git is complaining.

**Comment by timusus** [comment]
What do you mean 'git is complaining'? 

That file is committed, you can see it here:
https://github.com/timusus/Shuttle/blob/dev/.idea/codeStyles/codeStyleConfig.xml

**Comment by timusus** [comment]
@fxsalazar it's probably easier to have this discussion in Slack.
